### DISCUSSION TITLE
initrd flashing support

### DISCUSSION THREAD

**Discussion by akhilveeraghanta** [OE4T/meta-tegra#1115] [Q&A]
It looks like the meta-tegra repo doesn't copy over the initrd flashing scripts from the L4T source (in the tegra-binaries recipe?)

Is there a way to use the nvidia provided initrd flashing scripts with the output of a yocto/meta-tegra build? 

Also just curious if there is Orin NX support around the corner? (it exclusively uses the initrd flashing scripts, no emmc)

Thanks!
Akhil

**Comment by dwalkes** [comment]
@akhilveeraghanta we discussed some ideas for installing to NVMe in the meeting at https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2022-10-13 but I don't believe anyone has looked into this yet.

**Reply by akhilveeraghanta** [reply]
Flashing exclusively onto an nvme drive (w/out any emmc) is something that we are looking for to start using the Orin NX with our yocto stack. 

I would be interested in looking into/contributing to setting up installing to nvme, what would be the best way to go about this?

Is there a stop gap solution here that uses the scripts NVIDIA provides? (the `kernel_flash/l4t_initrd_flash.sh` script)

**Reply by dwalkes** [reply]
See related discussion at https://github.com/OE4T/meta-tegra/discussions/957#discussioncomment-2568383 - I don't think the NVIDIA scripts will work because I believe they require the image which runs from an initramfs, formats the NVME drive, and enables USB device mode to serve the NVME partitions.  I don't have experience with these myself, however.

**Reply by akhilveeraghanta** [reply]
They have additional scripts that do pretty much what you described in `kernel_flash/l4t_initrd*` that are not the standard tergra flash stuff over recovery + otg. 

They support booting into the recovery image, mounting the nvme drive over USB and flashing in the latest release. We could maybe steal some of that? 

Who would be a good person to talk to about implementing this feature?

**Reply by dwalkes** [reply]
> They support booting into the recovery image, mounting the nvme drive over USB and flashing in the latest release. We could maybe steal some of that?

Yes, I think one approach is to copy or attempt to reuse what they did, taking this reference from their readme:
```
Flash an NVMe SSD and use APP partition on it as root filesystem
sudo ./tools/kernel_flash/l4t_initrd_flash.sh --external-device nvme0n1p1 -c ./tools/kernel_flash/flash_l4t_nvme.xml -S 8GiB  --showlogs  jetson-xavier nvme0n1p1
```
Replacing the flash_l4t_nvme.xml reference with one you build based on the output content of a meta-tegra build, and iterating on that until it works.

A follow-on step is to automate this process, generating the xml files and the initrd as a part of the yocto build process, possibly leveraging some of the `initrd_flash` script content in the `kernel_flash` directory. This is what is suggested in https://github.com/OE4T/meta-tegra/discussions/957#discussioncomment-2568383

An alternative way is to do something different as described in https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2022-10-13  which does the repartitioning and partition writing step as a part of the initrd rather than passing through to the host.  This has some benefit in secure boot scenarios as discussed there and in other threads.  I think if you wanted to start on this the way I would suggest starting is to look at what we've done in the past for OTA reflashing as discussed in the meeting [here](https://github.com/OE4T/meta-tegra/wiki/OE4T-Meeting-Notes-2021-06-10#ota-re-flashing) and referenced discussions at https://github.com/OE4T/tegra-demo-distro/pull/113 and https://github.com/OE4T/tegra-demo-distro/discussions/117.  A benefit of this approach is that we'll need something similar to solve upgrade issues from Jetpack 4, so there will probably be lots of use cases which will benefit from this implementation.

> Who would be a good person to talk to about implementing this feature?

I think your best bet is to post progress or ideas here or in similar posts, see if others are interested in this.  You can also use [gitter](https://gitter.im/OE4T) for more interactive chat, or attend one of the [Monthly Meetings](https://github.com/OE4T/meta-tegra/discussions/515) and add to the agenda at the start of the call.

**Reply by akhilveeraghanta** [reply]
Thank you for the detailed response! I will take a closer look and report back

**Reply by mmitchel** [reply]
the kernel initrd flash scripts support "--massflash" which creates a tar file ; the dependency there requires the appropriate environment variables BOARDID CHIPREV FUSELEVEL BOARDSKU BOARDREV to generate that tar file without a target device connected by USB.  there is a README in the kernel_initrd folder that describes different use cases. 

one of the use cases is to generate a disk image that can be put onto either USB or NVME disks using dd or your favorite image writer. i have an unaddressed use case for ROOTFS_ENC=1 as part of the environment, im not sure meta-tegra has support for this even on the emmc.

**Comment by madisongh** [comment]
I've started some work in this area.  Wiki page is [here](https://github.com/OE4T/meta-tegra/wiki/initrd-flashing-support) and discussion, suggestions, etc. are welcome in the [Gitter room](https://gitter.im/OE4T/initrd-flashing).

**Reply by madisongh** [reply]
If you're using one of the JetPack 5/L4T R35.1.0-based branches, there is support there now for initrd-based flashing, including having support for an APP partition that resides on NVMe or USB storage.

**Reply by madisongh** [reply]
initrd flashing is now also in the `kirkstone-l4t-r32.7.x` branch.

**Reply by ichergui** [reply]
Excellent :smile:
